src/outputmodules/sacn.rs

```diff
@@ -0,0 +1,199 @@
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+use sacn::packet::ACN_SDT_MULTICAST_PORT;
+use sacn::source::SacnSource;
+
+use crate::{beatkeeper::TrackInfo, config::Config, log::ScopedLogger};
+use super::ModuleCreateOutput;
+use super::OutputModule;
+
+/// sACN (E1.31) output module
+///
+/// Config keys (with defaults):
+/// - `source` (String): local bind address, e.g. "0.0.0.0:5569". Default: bind to 0.0.0.0 on ACN port+1 (5569).
+/// - `mode` (String): "multicast" (default) or "unicast".
+/// - `universe` (u16): sACN universe (1..=63999), default 1.
+/// - `start_channel` (u16): DMX start/offset (1..=511), default 1. (We need 2 slots: beat count and BPM.)
+/// - `targets` (String): comma-separated IPv4 list for unicast. Example: "192.168.0.50,192.168.0.51".
+/// - `priority` (u8): sACN priority 1..200, default 100.
+/// - `source_name` (String): up to 63 ASCII chars shown by receivers. Default: "rkbx link".
+///
+/// Slot mapping (starting at `start_channel`):
+/// - +0 : BPM (u8). Capped to 250. Values > 250 are sent as 250.
+/// - +1 : Beat absolute counter (u8). Wraps 0..=255.
+///
+
+pub struct SACN {
+    src: SacnSource,
+    mode: Mode,
+    targets: Vec<SocketAddr>,
+    universe: u16,
+    start_slot: usize, // 1..=511 (we need 2 slots)
+    priority: u8,
+    local_addr: SocketAddr,
+    dmx: [u8; 513], // index 0 is start code = 0, then 512 DMX slots
+    logger: ScopedLogger,
+    last_beat_floor: i32,
+    beat_counter: u8,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum Mode { Multicast, Unicast }
+
+impl SACN 
+{
+    pub fn create(conf: Config, logger: ScopedLogger) -> ModuleCreateOutput {
+        // Local bind address
+        let source_name = conf.get_or_default("source_name", String::from("rkbx link"));
+        let bind_str: Option<String> = conf.get("source");
+        let local_addr = match bind_str {
+            Some(s) => {
+                // If the string contains a port, parse directly.
+                // Otherwise, let the system choose.
+                if s.contains(':') {
+                    s.parse::<SocketAddr>().expect("Invalid sACN bind addr")
+                } else {
+                    SocketAddr::new(
+                    s.parse::<IpAddr>().expect("Invalid sACN bind IP"),
+                    0,
+                    )
+                }
+            }
+            None => SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0),
+        };
+
+        let mut src = SacnSource::with_ip(&source_name, local_addr).expect("Failed to create SacnSource");
+
+        // Mode
+        let mode_str = conf.get_or_default("mode", String::from("multicast"));
+        let mode = match mode_str.to_ascii_lowercase().as_str() { "unicast" => Mode::Unicast, _ => Mode::Multicast };
```

Comment:
I'd like a warning when it defaults here, so

unicast-> Unicast,
multicast-> Multicast,
_ => Multicast + warning